{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Common settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Python Package Import\n",
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "##Ensuring dsub is up to date\n",
    "!pip3 install --upgrade dsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Setting for running dsub jobs\n",
    "pd.set_option('display.max_colwidth', 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Defining necessary pathways\n",
    "my_bucket = os.environ['WORKSPACE_BUCKET']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "USER_NAME = os.getenv('OWNER_EMAIL').split('@')[0].replace('.','-')\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env USER_NAME={USER_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#radix='antidiabetic.HbA1c'\n",
    "#radix='antidiabetic.HbA1cmyboxcox'\n",
    "#radix='antidiabetic.HbA1cmyboxcox01'\n",
    "#radix='antihypertensive.SBP'\n",
    "#radix='antihypertensive.SBPmyboxcox'\n",
    "#radix='antihypertensive.SBPmyboxcox01'\n",
    "#radix='antihypertensive.SBPmyboxcoxv2'\n",
    "#radix='hypolipidemics.LDL'\n",
    "#radix='hypolipidemics.LDLmyboxcox'\n",
    "#radix='hypolipidemics.LDLmyboxcox01'\n",
    "#radix='hypolipidemics.LDLdelta'\n",
    "#radix='hypolipidemics.LDLlogratio'\n",
    "#radix='hypolipidemics.LDLmyboxcoxv2'\n",
    "#radix='SBPnodrugmyboxcox-05nge4'\n",
    "#radix='MAPnodrugmyboxcox0nge4'\n",
    "#radix='PhysicalMeasurementsv2_MAPmyboxcoxnge2' # --phenoExcludeList measurement_datetime\n",
    "#radix='LabsMeasurementsv2_MAPmyboxcox10yrnge4'\n",
    "radix='LabsMeasurementsv2_MAPmyboxcox5yrnge4'\n",
    "#radix='PhysicalMeasurements_HRmyboxcoxnge2'# --phenoExcludeList measurement_datetime #suspended after step1\n",
    "#radix='LabsMeasurements_HRmyboxcox10yrnge4' #suspended after step1\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env radix={radix}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trait='qt' #bt\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env trait={trait}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Run below once per workspace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/aou_dsub.bash\n",
    "\n",
    "#!/bin/bash\n",
    "\n",
    "# This shell function passes reasonable defaults for several dsub parameters, while\n",
    "# allowing the caller to override any of them. It creates a nice folder structure within\n",
    "# the workspace bucket for dsub log files.\n",
    "\n",
    "# --[ Parameters ]--\n",
    "# any valid dsub parameter flag\n",
    "\n",
    "#--[ Returns ]--\n",
    "# the job id of the job created by dsub\n",
    "\n",
    "#--[ Details ]--\n",
    "# The first five parameters below should always be those values when running on AoU RWB.\n",
    "\n",
    "# Feel free to change the values for --user, --regions, --logging, and --image if you like.\n",
    "\n",
    "# Note that we insert some job data into the logging path.\n",
    "# https://github.com/DataBiosphere/dsub/blob/main/docs/logging.md#inserting-job-data\n",
    "\n",
    "function aou_dsub () {\n",
    "\n",
    "  # Get a shorter username to leave more characters for the job name.\n",
    "  local DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "  # For AoU RWB projects network name is \"network\".\n",
    "  #local AOU_NETWORK=network\n",
    "  #local AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "  dsub \\\n",
    "      --provider google-batch \\\n",
    "      --user-project \"${GOOGLE_PROJECT}\"\\\n",
    "      --project \"${GOOGLE_PROJECT}\"\\\n",
    "      --image 'ubuntu:latest' \\\n",
    "      --network \"global/networks/network\" \\\n",
    "      --subnetwork \"regions/us-central1/subnetworks/subnetwork\" \\\n",
    "      --service-account \"$(gcloud config get-value account)\" \\\n",
    "      --use-private-address \\\n",
    "      --user \"${DSUB_USER_NAME}\" \\\n",
    "      --regions us-central1 \\\n",
    "      --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "      \"$@\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "echo source ~/aou_dsub.bash >> ~/.bashrc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# dsub for regenie GWAS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## Prepare LD pruned all chr for regenie step 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "for chromo in {1..22}\n",
    "do\n",
    "echo $chromo\n",
    "\n",
    "gsutil -u $GOOGLE_PROJECT -m cp gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.pgen .\n",
    "gsutil -u $GOOGLE_PROJECT -m cp gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.psam .\n",
    "gsutil -u $GOOGLE_PROJECT -m cp gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.pvar .\n",
    "\n",
    "\n",
    "plink2 --pfile acaf_threshold.chr${chromo} \\\n",
    "  --maf 0.01 --mac 100 --geno 0.1 --hwe 1e-5 0.001 \\\n",
    "  --mind 0.1 \\\n",
    "  --write-snplist --write-samples --no-id-header \\\n",
    "  --set-missing-var-ids @:# \\\n",
    "  --out qc_pass\n",
    "\n",
    "plink2 --pfile acaf_threshold.chr${chromo} \\\n",
    "  --extract qc_pass.snplist \\\n",
    "  --keep qc_pass.id \\\n",
    "  --indep-pairwise 200kb 1 0.5 \\\n",
    "  --set-missing-var-ids @:# \\\n",
    "  --out ldpruned_snplist\n",
    "  \n",
    "plink2 --pfile acaf_threshold.chr${chromo} \\\n",
    "  --extract ldpruned_snplist.prune.in \\\n",
    "  --keep qc_pass.id \\\n",
    "  --make-pgen \\\n",
    "  --set-missing-var-ids @:# \\\n",
    "  --out acaf_threshold.ldpruned.chr${chromo}\n",
    "\n",
    "rm acaf_threshold.chr${chromo}.pgen\n",
    "rm acaf_threshold.chr${chromo}.psam\n",
    "rm acaf_threshold.chr${chromo}.pvar\n",
    "rm qc_pass.*\n",
    "rm ldpruned_snplist.*\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "`--pmerge-list` does not work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "ln -s acaf_threshold.ldpruned.chr1.pgen foo.1.pgen\n",
    "ln -s acaf_threshold.ldpruned.chr1.psam foo.1.psam\n",
    "ln -s acaf_threshold.ldpruned.chr1.pvar foo.1.pvar\n",
    "\n",
    "for c in {1..21}\n",
    "do\n",
    "    d=$((c+1))\n",
    "    plink2 --pfile foo.$c --pmerge acaf_threshold.ldpruned.chr$d --mind 0.1 \\\n",
    "      --make-pgen --out foo.$d\n",
    "    rm foo.$c.{pgen,psam,pvar} foo.${d}-merge.{pgen,psam,pvar}\n",
    "done\n",
    "\n",
    "mv foo.22.pgen acaf_threshold.ldpruned.chrall.pgen\n",
    "mv foo.22.psam acaf_threshold.ldpruned.chrall.psam\n",
    "mv foo.22.pvar acaf_threshold.ldpruned.chrall.pvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "plink2 --pfile acaf_threshold.ldpruned.chrall --missing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "plink2 --pfile acaf_threshold.ldpruned.chrall \\\n",
    "    --max-alleles 2 \\\n",
    "    --export bgen-1.3 \\\n",
    "    --out acaf_threshold.ldpruned.chrall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%% bash\n",
    "\n",
    "mv acaf_threshold.ldpruned.chrall.sample acaf_threshold.ldpruned.chrall.sample.original\n",
    "perl -ne 's/^0 (\\d+)/$1 $1/; print' < acaf_threshold.ldpruned.chrall.sample.original > acaf_threshold.ldpruned.chrall.sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "`gcloud storage cp` does not work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil cp acaf_threshold.ldpruned.chrall.bgen $my_bucket\"/data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil cp acaf_threshold.ldpruned.chrall.sample $my_bucket\"/data/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## regenie step 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "! grep -v NA < aou_{radix}_covariates.txt > qc_pass.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "plink2 --bgen acaf_threshold.ldpruned.chrall.bgen ref-unknown \\\n",
    "  --sample acaf_threshold.ldpruned.chrall.sample \\\n",
    "  --maf 0.01 --mac 100 --geno 0.1 --hwe 1e-5 0.001 \\\n",
    "  --write-snplist -no-id-header \\\n",
    "  --keep qc_pass.id \\\n",
    "  --out qc_pass  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "! gsutil cp qc_pass.snplist {my_bucket}/data/aou_{radix}.qc_pass.snplist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## MODIFY FOR FULL DATA RUN \n",
    "# Use underscores, not whitespace since it will become part of the bucket path.\n",
    "JOB_NAME=f'regenie_{radix}' ## add name in quotes, copy name in quotes to 4.1\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_NAME={JOB_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Analysis Results Folder \n",
    "line_count_results_folder = os.path.join(\n",
    "    os.getenv('WORKSPACE_BUCKET'),\n",
    "    'dsub',\n",
    "    'results',\n",
    "    JOB_NAME,\n",
    "    USER_NAME,\n",
    "    datetime.now().strftime('%Y%m%d'))\n",
    "\n",
    "line_count_results_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Where the output files will go\n",
    "output_files = os.path.join(line_count_results_folder, \"results\")\n",
    "print(output_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "OUTPUT_FILES = output_files\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env OUTPUT_FILES={OUTPUT_FILES}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/Regenie_GWAS_NewPlinkBgens.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "cp \"${regenie_file}\" regenie.zip\n",
    "unzip regenie.zip\n",
    "mv regenie_v4.1.gz_x86_64_Linux regenie\n",
    "chmod 700 regenie\n",
    "\n",
    "./regenie \\\n",
    "    --step 1 \\\n",
    "    --bgen \"${bgen_file}\" \\\n",
    "    --sample \"${sample_file}\" \\\n",
    "    --phenoFile \"${pheno_file}\" --phenoExcludeList measurement_datetime \\\n",
    "    --covarFile \"${cov_file}\" \\\n",
    "    --extract \"${snplist_file}\" \\\n",
    "    --\"${trait}\" \\\n",
    "    --bsize 1000 \\\n",
    "    --lowmem \\\n",
    "    --lowmem-prefix regenie_tmp_preds \\\n",
    "    --verbose \\\n",
    "    --out \"${prefix}\"_step1_out\n",
    "\n",
    "zip \"${prefix}\"_step1.zip \"${prefix}\"_step1_out*\n",
    "mv \"${prefix}\"_step1.zip \"${OUTPUT_PATH}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/Regenie_GWAS_NewPlinkBgens.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil ls {my_bucket}/data/dsub/*.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The next cell contains the dsub command that will kick off the Regenie GWAS once ran. Right now, the ```--input``` variable options below point to the paths of the input files used in the original v7 LDL-C Regenie GWAS. If you would like to re-create the original run of the v7 LDL GWAS, change nothing (other than the lower and upper bounds depending on your test region) and run the next cell as is.\n",
    "\n",
    "If you just created new phenotype and covariate files using notebooks 1-3 and section 2 above, you will need to change the ```--input pheno_file``` and ```--input cov_file``` variable paths to point to your new files in your bucket. \n",
    "\n",
    "If you just ran sections 3-6 and have new bgen files, you will need to change the ```--input bgen_file``` and ```--input sample_file``` variable paths to point to your new bgen files in your bucket."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "2026.10.24 invalid: --image \"gcr.io/bick-aps2/ghcr.io/rgcgithub/regenie/regenie:v3.2.4.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "! gcloud container images describe \"gcr.io/bick-aps2/ghcr.io/rgcgithub/regenie/regenie:v3.2.4.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "! gcloud container images describe \"us.gcr.io/broad-dsp-gcr-public/terra-jupyter-aou:2.2.16\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "NOT CHECKED\n",
    "We would also like to share that you can search docker hub to find one image, e.g., --image shengqh/regenie4:20241127, and we recommend running a small test first before using such images."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "%%bash --out LINE_COUNT_JOB_ID\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "# Get a shorter username to leave more characters for the job name.\n",
    "DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "    aou_dsub \\\n",
    "    --name \"${JOB_NAME}\" \\\n",
    "    --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "    --image \"us.gcr.io/broad-dsp-gcr-public/terra-jupyter-aou:2.2.16\" \\\n",
    "    --machine-type \"n2-standard-4\" \\\n",
    "    --boot-disk-size 1000 \\\n",
    "    --script \"${WORKSPACE_BUCKET}/data/dsub/Regenie_GWAS_NewPlinkBgens.sh\" \\\n",
    "    --input bgen_file=\"${WORKSPACE_BUCKET}/data/acaf_threshold.ldpruned.chrall.bgen\" \\\n",
    "    --input sample_file=\"${WORKSPACE_BUCKET}/data/acaf_threshold.ldpruned.chrall.sample\" \\\n",
    "    --input pheno_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}_QT.txt\" \\\n",
    "    --input cov_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}_covariates.txt\" \\\n",
    "    --input snplist_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}.qc_pass.snplist\" \\\n",
    "    --input regenie_file=\"${WORKSPACE_BUCKET}/data/regenie_v4.1.gz_x86_64_Linux.zip\" \\\n",
    "    --output-recursive OUTPUT_PATH=\"${OUTPUT_FILES}\" \\\n",
    "    --env prefix=\"aou_${radix}\" \\\n",
    "    --env trait=${trait} \\\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Save this Python variable value as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_ID={LINE_COUNT_JOB_ID}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "dstat \\\n",
    "    --provider google-batch \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --jobs \"${JOB_ID}\" \\\n",
    "    --users \"${USER_NAME}\" \\\n",
    "    --status '*' --format json --full"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "## regenie step 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## MODIFY FOR FULL DATA RUN \n",
    "# Use underscores, not whitespace since it will become part of the bucket path.\n",
    "JOB_NAME=f'regenie_{radix}' ## add name in quotes, copy name in quotes to 4.1\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_NAME={JOB_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Analysis Results Folder \n",
    "line_count_results_folder = os.path.join(\n",
    "    os.getenv('WORKSPACE_BUCKET'),\n",
    "    'dsub',\n",
    "    'results',\n",
    "    JOB_NAME,\n",
    "    USER_NAME,\n",
    "    datetime.now().strftime('%Y%m%d'))\n",
    "\n",
    "line_count_results_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Where the output files will go\n",
    "output_files = os.path.join(line_count_results_folder, \"results\")\n",
    "print(output_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "OUTPUT_FILES = output_files\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env OUTPUT_FILES={OUTPUT_FILES}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/Regenie_GWAS_NewPlinkBgens2.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "cp \"${regenie_file}\" regenie.zip\n",
    "unzip regenie.zip\n",
    "mv regenie_v4.1.gz_x86_64_Linux regenie\n",
    "chmod 700 regenie\n",
    "\n",
    "cp \"${step1zip_file}\" foo.zip\n",
    "unzip foo.zip\n",
    "\n",
    "./regenie \\\n",
    "    --step 2 \\\n",
    "    --bgen   \"${bgen_file}\" \\\n",
    "    --sample \"${sample_file}\" \\\n",
    "    --phenoFile \"${pheno_file}\" --phenoExcludeList measurement_datetime \\\n",
    "    --covarFile \"${cov_file}\" \\\n",
    "    --\"${trait}\" \\\n",
    "    --firth --approx --pThresh 0.01 \\\n",
    "    --pred \"${prefix}\"_step1_out_pred.list \\\n",
    "    --bsize 400 \\\n",
    "    --verbose \\\n",
    "    --ref-first \\\n",
    "    --out \"${prefix}\"_step2_out_chr\"${chrom}\"\n",
    "\n",
    "mv \"${prefix}\"_step2_out_chr\"${chrom}\"* ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/Regenie_GWAS_NewPlinkBgens2.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil ls {my_bucket}/data/dsub/*.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The next cell contains the dsub command that will kick off the Regenie GWAS once ran. Right now, the ```--input``` variable options below point to the paths of the input files used in the original v7 LDL-C Regenie GWAS. If you would like to re-create the original run of the v7 LDL GWAS, change nothing (other than the lower and upper bounds depending on your test region) and run the next cell as is.\n",
    "\n",
    "If you just created new phenotype and covariate files using notebooks 1-3 and section 2 above, you will need to change the ```--input pheno_file``` and ```--input cov_file``` variable paths to point to your new files in your bucket. \n",
    "\n",
    "If you just ran sections 3-6 and have new bgen files, you will need to change the ```--input bgen_file``` and ```--input sample_file``` variable paths to point to your new bgen files in your bucket."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%bash --out LINE_COUNT_JOB_ID\n",
    "\n",
    "# Get a shorter username to leave more characters for the job name.\n",
    "DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "# Python is 'right side limited' wherein the last value is not included\n",
    "# To run the regression across all chromosomes, set lower to 1 and upper to 23\n",
    "# To run across one chromosome, set lower to the chomosome-of-interest and upper to the following\n",
    "\n",
    "LOWER=1\n",
    "UPPER=23\n",
    "for ((chromo=$LOWER;chromo<$UPPER;chromo+=1))\n",
    "do\n",
    "\n",
    "    aou_dsub \\\n",
    "    --name \"${JOB_NAME}\" \\\n",
    "    --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "    --image \"us.gcr.io/broad-dsp-gcr-public/terra-jupyter-aou:2.2.16\" \\\n",
    "    --machine-type \"n2-standard-4\" \\\n",
    "    --boot-disk-size 1000 \\\n",
    "    --disk-size 1000 \\\n",
    "    --script \"${WORKSPACE_BUCKET}/data/dsub/Regenie_GWAS_NewPlinkBgens2.sh\" \\\n",
    "    --input regenie_file=\"${WORKSPACE_BUCKET}/data/regenie_v4.1.gz_x86_64_Linux.zip\" \\\n",
    "    --input step1zip_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}_step1.zip\" \\\n",
    "    --input bgen_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.bgen\" \\\n",
    "    --input sample_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.sample\" \\\n",
    "    --input pheno_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}_QT.txt\" \\\n",
    "    --input cov_file=\"${WORKSPACE_BUCKET}/data/aou_${radix}_covariates.txt\" \\\n",
    "    --output-recursive OUTPUT_PATH=\"${OUTPUT_FILES}/${chromo}\" \\\n",
    "    --env prefix=\"aou_${radix}\" \\\n",
    "    --env trait=${trait} \\\n",
    "    --env chrom=${chromo} \\\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Save this Python variable value as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_ID={LINE_COUNT_JOB_ID}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "dstat \\\n",
    "    --provider google-batch \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --jobs \"${JOB_ID}\" \\\n",
    "    --users \"${USER_NAME}\" \\\n",
    "    --status '*'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# dsub for plink hardy, GWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## MODIFY FOR FULL DATA RUN \n",
    "# Use underscores, not whitespace since it will become part of the bucket path.\n",
    "JOB_NAME=f'plink_{radix}' ## add name in quotes, copy name in quotes to 4.1\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_NAME={JOB_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Analysis Results Folder \n",
    "line_count_results_folder = os.path.join(\n",
    "    os.getenv('WORKSPACE_BUCKET'),\n",
    "    'dsub',\n",
    "    'results',\n",
    "    JOB_NAME,\n",
    "    USER_NAME,\n",
    "    datetime.now().strftime('%Y%m%d'))\n",
    "\n",
    "line_count_results_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Where the output files will go\n",
    "output_files = os.path.join(line_count_results_folder, \"results\")\n",
    "print(output_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "OUTPUT_FILES = output_files\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env OUTPUT_FILES={OUTPUT_FILES}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/hardy.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "plink2 \\\n",
    "  --bgen ${bgen_file} ref-unknown \\\n",
    "  --keep ${keep_file} \\\n",
    "  --hardy \\\n",
    "  --out chr${chrom}\n",
    "\n",
    "mv chr${chrom}.hardy ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/hardy.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%bash --out LINE_COUNT_JOB_ID\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "# Get a shorter username to leave more characters for the job name.\n",
    "DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "LOWER=1\n",
    "UPPER=23\n",
    "for ((chromo=$LOWER;chromo<$UPPER;chromo+=1))\n",
    "do\n",
    "\n",
    "    aou_dsub \\\n",
    "    --name \"${JOB_NAME}\" \\\n",
    "    --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "    --image \"us.gcr.io/broad-dsp-gcr-public/terra-jupyter-aou:2.2.16\" \\\n",
    "    --machine-type \"n2-standard-4\" \\\n",
    "    --boot-disk-size 1000 \\\n",
    "    --disk-size 1000 \\\n",
    "    --script \"${WORKSPACE_BUCKET}/data/dsub/hardy.sh\" \\\n",
    "    --output-recursive OUTPUT_PATH=\"${OUTPUT_FILES}/${chromo}\" \\\n",
    "    --input bgen_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.bgen\" \\\n",
    "    --input sample_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.sample\" \\\n",
    "    --input keep_file=\"${WORKSPACE_BUCKET}/data/eur.plink.ind\" \\\n",
    "    --env chrom=${chromo} \\\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!dstat --provider google-batch --project terra-vpc-sc-9b3acbf3 --location us-central1 --jobs 'plink-labs--fumibaker--251029-203511-73' --users 'fumibaker' --status '*'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "variance inflation factor for covariate 'PC6' is too high (VIF_TOO_HIGH).\n",
    "==> limit up to PC5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/plinkGWAS.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "cp \"${plink2_file}\" ./plink2\n",
    "chmod 700 ./plink2\n",
    "\n",
    "./plink2 \\\n",
    "  --pgen ${pgen_file} \\\n",
    "  --pvar ${pvar_file} \\\n",
    "  --psam ${psam_file} \\\n",
    "  --glm \\\n",
    "  --out ${trait}.chr${chrom} \\\n",
    "  --pheno \"${pheno_file}\" \\\n",
    "  --pheno-name ${trait} \\\n",
    "  --covar \"${cov_file}\" \\\n",
    "  --covar-name PC1 PC2 PC3 PC4 PC5 \\\n",
    "  --covar-variance-standardize \\\n",
    "  --maf 0.01\n",
    "\n",
    "mv \"${trait}.chr${chrom}\"* ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/plinkGWAS.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil ls {my_bucket}/data/dsub/*.sh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "!gsutil -u $GOOGLE_PROJECT ls gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%bash --out LINE_COUNT_JOB_ID\n",
    "\n",
    "# Get a shorter username to leave more characters for the job name.\n",
    "DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "# For AoU RWB projects network name is \"network\".\n",
    "AOU_NETWORK=network\n",
    "AOU_SUBNETWORK=subnetwork\n",
    "\n",
    "MACHINE_TYPE=\"n2-standard-4\"\n",
    "\n",
    "# Change for your bucket, path in output of cell directly above:\n",
    "BASH_SCRIPT=\"gs://fc-secure-7ac6cae0-10b6-47a1-acb4-c71d46ca046e/data/dsub/plinkGWAS.sh\"\n",
    "\n",
    "#LOWER=1\n",
    "#UPPER=22\n",
    "#for ((chromo=$LOWER;chromo<$UPPER;chromo+=1))\n",
    "for chromo in 22\n",
    "do\n",
    "#for t in C10AA C10AB C10AC C10AD C10AX06 C10AX09\n",
    "for t in C10AB\n",
    "do\n",
    "    dsub \\\n",
    "    --provider google-cls-v2 \\\n",
    "    --user-project \"${GOOGLE_PROJECT}\" \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --image \"gcr.io/bick-aps2/ghcr.io/rgcgithub/regenie/regenie:v3.2.4.gz\" \\\n",
    "    --network \"${AOU_NETWORK}\" \\\n",
    "    --subnetwork \"${AOU_SUBNETWORK}\" \\\n",
    "    --service-account \"$(gcloud config get-value account)\" \\\n",
    "    --user \"${DSUB_USER_NAME}\" \\\n",
    "    --regions us-central1 \\\n",
    "    --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "    \"$@\" \\\n",
    "    --preemptible \\\n",
    "    --boot-disk-size 1000 \\\n",
    "    --machine-type ${MACHINE_TYPE} \\\n",
    "    --name \"${JOB_NAME}\" \\\n",
    "    --script \"${BASH_SCRIPT}\" \\\n",
    "    --env GOOGLE_PROJECT=${GOOGLE_PROJECT} \\\n",
    "    --input plink2_file=\"gs://fc-secure-6eb7a615-fd07-4e0e-9394-928a16db191c/data/plink2\" \\\n",
    "    --input pgen_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.pgen\" \\\n",
    "    --input pvar_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.pvar\" \\\n",
    "    --input psam_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/pgen/acaf_threshold.chr${chromo}.psam\" \\\n",
    "    --input pheno_file=\"gs://fc-secure-7ac6cae0-10b6-47a1-acb4-c71d46ca046e/data/aou_hypolipidemics.LDL_QT.txt\" \\\n",
    "    --input cov_file=\"gs://fc-secure-7ac6cae0-10b6-47a1-acb4-c71d46ca046e/data/aou_hypolipidemics.LDL_covariates.txt\" \\\n",
    "    --env trait=${t} \\\n",
    "    --env chrom=${chromo} \\\n",
    "    --output-recursive OUTPUT_PATH=\"${OUTPUT_FILES}/${chromo}\"\n",
    "done\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# dsub for plink PGS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## MODIFY FOR FULL DATA RUN \n",
    "# Use underscores, not whitespace since it will become part of the bucket path.\n",
    "#JOB_NAME='regenie_hypertensiondrug' ## add name in quotes, copy name in quotes to 4.1\n",
    "JOB_NAME='score_lipidaemiadrug' ## add name in quotes, copy name in quotes to 4.1\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_NAME={JOB_NAME}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Analysis Results Folder \n",
    "line_count_results_folder = os.path.join(\n",
    "    os.getenv('WORKSPACE_BUCKET'),\n",
    "    'dsub',\n",
    "    'results',\n",
    "    JOB_NAME,\n",
    "    USER_NAME,\n",
    "    datetime.now().strftime('%Y%m%d'))\n",
    "\n",
    "line_count_results_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "## Where the output files will go\n",
    "output_files = os.path.join(line_count_results_folder, \"results\")\n",
    "print(output_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "OUTPUT_FILES = output_files\n",
    "\n",
    "# Save this Python variable as an environment variable so that its easier to use within %%bash cells.\n",
    "%env OUTPUT_FILES={OUTPUT_FILES}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "2025.09.29 added option --score cols=+scoresums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/score.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "cp \"${plink2_file}\" ./plink2\n",
    "chmod 700 ./plink2\n",
    "\n",
    "cat ${sample_file} | \\\n",
    "  sed \"s/^0\\t0\\t0\\.0/0\\t0\\t0/\" | \\\n",
    "  cut -f 1,2,3 > mysample\n",
    "\n",
    "./plink2 \\\n",
    "  --bgen ${bgen_file} ref-unknown \\\n",
    "  --sample mysample \\\n",
    "  --score ${coeff_file} list-variants cols=+scoresums \\\n",
    "  --out score.${trait}.chr${chrom}\n",
    "\n",
    "mv \"score.${trait}.chr${chrom}\"* ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%writefile ~/export.sh\n",
    "\n",
    "set -o pipefail \n",
    "set -o errexit\n",
    "\n",
    "cp \"${plink2_file}\" ./plink2\n",
    "chmod 700 ./plink2\n",
    "\n",
    "./plink2 \\\n",
    "  --bgen ${bgen_file} ref-unknown \\\n",
    "  --snp chr10:38839760:G:C \\\n",
    "  --export A \\\n",
    "  --out chr10:38839760:G:C\n",
    "\n",
    "mv chr10:38839760:G:C* ${OUTPUT_PATH}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/score.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": [],
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## Modify to personal gs bucket\n",
    "!gsutil cp /home/jupyter/export.sh {my_bucket}/data/dsub/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "!gsutil ls {my_bucket}/data/dsub/*.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "hidden": true
   },
   "source": [
    "The next cell contains the dsub command that will kick off the Regenie GWAS once ran. Right now, the ```--input``` variable options below point to the paths of the input files used in the original v7 LDL-C Regenie GWAS. If you would like to re-create the original run of the v7 LDL GWAS, change nothing (other than the lower and upper bounds depending on your test region) and run the next cell as is.\n",
    "\n",
    "If you just created new phenotype and covariate files using notebooks 1-3 and section 2 above, you will need to change the ```--input pheno_file``` and ```--input cov_file``` variable paths to point to your new files in your bucket. \n",
    "\n",
    "If you just ran sections 3-6 and have new bgen files, you will need to change the ```--input bgen_file``` and ```--input sample_file``` variable paths to point to your new bgen files in your bucket."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "%%bash --out LINE_COUNT_JOB_ID\n",
    "\n",
    "source ~/aou_dsub.bash # This file was created via notebook 01_dsub_setup.ipynb.\n",
    "\n",
    "# Get a shorter username to leave more characters for the job name.\n",
    "DSUB_USER_NAME=\"$(echo \"${OWNER_EMAIL}\" | cut -d@ -f1)\"\n",
    "\n",
    "LOWER=1\n",
    "UPPER=23\n",
    "for ((chromo=$LOWER;chromo<$UPPER;chromo+=1))\n",
    "do\n",
    "#for t in C10AA C10AB C10AX09\n",
    "for t in C03 C07 C08 C09\n",
    "#for t in C10AX09\n",
    "#for t in PA\n",
    "do\n",
    "\n",
    "\n",
    "    aou_dsub \\\n",
    "    --name \"${JOB_NAME}\" \\\n",
    "    --logging \"${WORKSPACE_BUCKET}/dsub/logs/{job-name}/{user-id}/$(date +'%Y%m%d/%H%M%S')/{job-id}-{task-id}-{task-attempt}.log\" \\\n",
    "    --image \"gcr.io/bick-aps2/ghcr.io/rgcgithub/regenie/regenie:v3.2.4.gz\" \\\n",
    "    --machine-type \"n2-standard-4\" \\\n",
    "    --boot-disk-size 1000 \\\n",
    "    --disk-size 1000 \\\n",
    "    --script \"${WORKSPACE_BUCKET}/data/dsub/score.sh\" \\\n",
    "    --output-recursive OUTPUT_PATH=\"${OUTPUT_FILES}/${chromo}\" \\\n",
    "    --input plink2_file=\"gs://fc-secure-6eb7a615-fd07-4e0e-9394-928a16db191c/data/plink2\" \\\n",
    "    --input bgen_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.bgen\" \\\n",
    "    --input sample_file=\"gs://fc-aou-datasets-controlled/v8/wgs/short_read/snpindel/acaf_threshold/bgen/chr${chromo}.sample\" \\\n",
    "    --input coeff_file=\"${WORKSPACE_BUCKET}/data/coeff.${t}.hg38.txt\" \\\n",
    "    --env trait=${t} \\\n",
    "    --env chrom=${chromo} \\\n",
    "\n",
    "done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "# Save this Python variable value as an environment variable so that its easier to use within %%bash cells.\n",
    "%env JOB_ID={LINE_COUNT_JOB_ID}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "! gsutil cat gs://fc-secure-7ac6cae0-10b6-47a1-acb4-c71d46ca046e/dsub/logs/score-lipidaemiadrug/fumibaker/20250912/043440/score-lipi--fumibaker--250912-043442-45-task-None-stdout.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "dstat \\\n",
    "    --provider google-batch \\\n",
    "    --project \"${GOOGLE_PROJECT}\" \\\n",
    "    --location us-central1 \\\n",
    "    --jobs \"${JOB_ID}\" \\\n",
    "    --users \"${USER_NAME}\" \\\n",
    "    --status '*'"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {
    "height": "calc(100% - 180px)",
    "left": "10px",
    "top": "150px",
    "width": "165px"
   },
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
